### DISCUSSION TITLE
Half-parsed models

### DISCUSSION THREAD

**Discussion by goto40** [textX/textX#368] [General]
I am thinking if/how we could achieve half-parsed models (incomplete models, e.g, textually available up to a point where a reference is expected). This could be interesting for a LSP/editor support...

First thing is to allow arpeggio to return an incomplete parse tree in case of an error (+the error).

This is just an idea... Maybe we need to think about it and look how other parser frameworks achieve this task.

**Comment by igordejanovic** [comment]
Maybe it would be better to have error recovery where a recovered tree/model would contain error markers. This would enable both code completion (especially if referred thing is ahead of the current location) as well as reporting all errors in one go.
We had a related discussion previously on #202. I also remember that we discussed this in textX-LS project but can't find that discussion at the moment.

An interesting approach in error recovery is implemented by the [tree-sitter project](https://tree-sitter.github.io/tree-sitter/) although tree-sitter IIRC is based on GLR so not sure that those ideas are applicable to textX.
